### DISCUSSION TITLE
Im not able to run Detox test on an existing APK

### DISCUSSION THREAD

**Discussion by g4ss4n** [wix/Detox#4307] [Q&A]
Hi, 

My end goal is to run Detox tests on an existing APK. We are using React Native. 

When running the following command 
`detox test --configuration android.emu.release`

13:31:43.291 detox[42887] B jest --config detox/e2e/jest.config.js
13:31:44.601 detox[42888] i DetoxRuntimeError: Test APK at path /Users/ghassangharzuzy/Documents/GitHub/agwa-testing/detox/nativeApps/android/release/app.apk was detected as the *app* APK!

HINT: Your test test-binary path was probably wrongly set in the active Detox configuration.
For further assistance, visit the Android setup guide: https://github.com/wix/Detox/blob/master/docs/Introduction.Android.md
13:31:47.596 detox[42888] i planting.test.js is assigned to emulator-5554 (Nexus_5X_API_23)
13:31:47.689 detox[42888] i Planting flow: Can login to the app
13:31:47.690 detox[42888] i Planting flow: Can login to the app [FAIL]
 ....
    No instrumentation runner found on device emulator-5554 for package com.agwa.mobileapp

      38 | describe('Planting flow', () => {
      39 |   beforeAll(async () => {
    > 40 |     await device.launchApp({permissions: {notifications: 'YES'}});
         |                  ^
      41 |   });
      42 |

I am able to see that the app was installed in the emulator. 

I can see there is an example in a similar discussion here - but the example is no longer available - https://github.com/wix/Detox/issues/3188 

Is anyone familiar with this issue? Any idea how to fix it?

**Comment by noomorph** [comment]
binaryPath and testBinaryPath are different things, and each should lead to its own APK file.

The test APK is created by commands like `gradlew assembleAndroidTest ...` e.g.:

![image](https://github.com/wix/Detox/assets/1962469/21c104f5-a3ee-44f8-9265-bea2bd493f33)


Make sure to finish your project setup: https://wix.github.io/Detox/docs/introduction/project-setup

Normally you don't need to specify `testBinaryPath` separately, but if it is created in a custom location, then you should add it along with `binaryPath` in your Detox config.

**Reply by g4ss4n** [reply]
Just to be sure I understood this correctly - do I need to change anything in my app configuration before the build?

I have the APK in a different repo, where I drag the existing APK to the given location, and I can see that the command installs the app in the emulator ... 

Worth mentioning - Im not looking to interfere with Gradle / the build of the APK or with the code there, Im only taking an APK, moving it to the testing repo where I have detox and run my tests from there 

```
/** @type {Detox.DetoxConfig} */
module.exports = {
  testRunner: {
    args: {
      '$0': 'jest',
      config: 'detox/e2e/jest.config.js',
    },
    jest: {
      setupTimeout: 120000
    }
  },
  devices: {
    emulator: {
      type: 'android.emulator',
      device: {
        avdName: 'Nexus_5X_API_23'
      }
    },
    simulator: {
      type: 'ios.simulator',
      device: {
        type: 'iPhone 14 Plus'
      }
    }
  },
  apps: {
    'android.debug': {
      type: 'android.apk',
      binaryPath: './detox/nativeApps/android/debug/1.0.38(148)-debug.apk',
      testBinaryPath: './detox/nativeApps/android/androidTest/debug/1.0.38(148)-debug-androidTest.apk',
    },
    'android.release': {
      type: 'android.apk',
      binaryPath: './detox/nativeApps/android/release/app.apk',
      testBinaryPath: './detox/nativeApps/android/release/app.apk'
    },
    'ios.debug': {
      type: 'ios.app',
      binaryPath: './detox/nativeApps/ios/Agwa.app'
    }
  },
  configurations: {
    'android.emu.debug': {
      device: 'emulator',
      app: 'android.debug'
    },
    'android.emu.release': {
      device: 'emulator',
      app: 'android.release'
    },
    'ios.sim.debug': {
      device: 'simulator',
      app: 'ios.debug'
    },
    'ios.sim.release': {
      device: 'simulator',
      app: 'ios.release' 
    }
  }
};
```


![image](https://github.com/wix/Detox/assets/49415228/af102d4c-4aba-4724-92df-18c9ed4a232e)

**Reply by d4vidi** [reply]
@g4ss4n The configuration associated with `android.emu.release` (or more specifically, the `android.release` app) seems suspicious:

```js
    'android.release': {
      type: 'android.apk',
      binaryPath: './detox/nativeApps/android/release/app.apk',
      testBinaryPath: './detox/nativeApps/android/release/app.apk'
    }
```

The error suggests that the APK in the path for `testBinaryPath` is of the _app_ APK rather than the test APK. Indeed, this can be seen in the configuration, as `testBinaryPath` appears to be the same as `binaryPath` (which is the path for the app APK); All the more so, it doesn't lead to an APK file whose name is in the form of (or similar to) `app-debug-androidTest.apk`, which is typically the exported name for test APK's.

Your configuration for `debug` does seem proper, nevertheless:

```js
    'android.debug': {
      type: 'android.apk',
      binaryPath: './detox/nativeApps/android/debug/1.0.38(148)-debug.apk',
      testBinaryPath: './detox/nativeApps/android/androidTest/debug/1.0.38(148)-debug-androidTest.apk',
    }
```

I think you ought to use the exact same `testBinaryPath` for the `android.release` config as well (it's recommended to use test APK compiled in Debug mode), namely:

```js
    'android.release': {
      type: 'android.apk',
      binaryPath: './detox/nativeApps/android/release/app.apk',
      testBinaryPath: './detox/nativeApps/android/androidTest/debug/1.0.38(148)-debug-androidTest.apk',
    },
```
